### ISSUE TITLE
Absence of Date() Object in JavaScript

### ISSUE THREAD

**Issue by ASimnegar** [Khan/live-editor#520]
Currently, on Khan Academy, the `Date` object is disallowed with the error message:
```Oh noes!
undefined is not an object (evaluating 'e.prototype')```

There is **no reason whatsoever** to not have the `Date` object accessible in Khan Academy. Can we please have it back?

**Comment by pamelafox** [comment]
Thanks for filing! I believe the place to add it is here:
https://github.com/Khan/live-editor/blob/master/js/output/pjs/pjs-code-injector.js#L125
I don't know of a reason not to add it, so I'll send a PR with a test later this week
if someone doesn't beat me to it.

**Comment by kevinbarabash** [comment]
This one is a bit tricky because `Date.now()` is stateful.

**Comment by ASimnegar** [comment]
What does stateful mean? Isn't it like a getter?

On Tue, Dec 15, 2015 at 9:53 PM, Kevin Barabash <notifications@github.com>
wrote:

> This one is a bit tricky because Date.now() is stateful.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Khan/live-editor/issues/520#issuecomment-164970034>.
>



-- 
Sincerely,

Ariel Simnegar

**Comment by GoToLoop** [comment]
Stateful means it is an object w/ mutable instance properties.

**Comment by kevinbarabash** [comment]
Maybe `stateful` was the best choice of words.  The issue is that live-editor let's you change parts of your program while it's running.  This often requires re-running commands because processing-js doesn't use a scene graph and you need to redraw the scene each time.  The problem is that some functions are not idempotent.  This means that they do not produce the same result if you call them multiple times.  `random` is one example of these.  We solve this by keeping track of the random seed and resetting the seed whenever we have to re-run code.  

The problem with `Date.now` is that this approach doesn't make sense because if a user is printing the time they want to see the actual time.  At the same time, if `Date.now` is being used to position object A on the screen and the user is modifying a variable that controls object B, both objects will move even though the user is only updating one.  I'm not sure how confusing this will be for novice programmers.

All of that being said, we should probably just expose the `Date` object along with `Date.now` and see what kind of weird bugs fall out, if any.

**Comment by ASimnegar** [comment]
It doesn't seem to be working. When I try to create a Date, I get a
TypeError:

TypeError: Cannot read property 'prototype' of undefined

On Thu, Dec 17, 2015 at 9:13 PM, Kevin Barabash <notifications@github.com>
wrote:

> Maybe stateful was the best choice of words. The issue is that
> live-editor let's you change parts of your program while it's running. This
> often requires re-running commands because processing-js doesn't use a
> scene graph and you need to redraw the scene each time. The problem is that
> some functions are not idempotent. This means that they do not produce the
> same result if you call them multiple times. random is one example of
> these. We solve this by keeping track of the random seed and resetting the
> seed whenever we have to re-run code.
>
> The problem with Date.now is that this approach doesn't make sense
> because if a user is printing the time they want to see the actual time. At
> the same time, if Date.now is being used to position object A on the
> screen and the user is modifying a variable that controls object B, both
> objects will move even though the user is only updating one. I'm not sure
> how confusing this will be for novice programmers.
>
> All of that being said, we should probably just expose the Date object
> along with Date.now and see what kind of weird bugs fall out, if any.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Khan/live-editor/issues/520#issuecomment-165646725>.
>



-- 
Sincerely,

Ariel Simnegar

**Comment by kevinbarabash** [comment]
@ASimnegar are you trying to fix this issue?

**Comment by ASimnegar** [comment]
I don't really know much about Github, but I can test it

Sincerely,
   Ariel Simnegar

**Comment by DozenalFTW** [comment]
> The problem with Date.now is that this approach doesn't make sense
  because if a user is printing the time they want to see the actual time. At
 the same time, if Date.now is being used to position object A on the
 screen and the user is modifying a variable that controls object B, both
 objects will move even though the user is only updating one. I'm not sure
 how confusing this will be for novice programmers.

Uhhh... if they knew what `Date` did, they would know that they don't have the ability to freeze time.

**Comment by MatthiasSaihttam** [comment]
Can be closed as #585 has been merged.
